Add `venv`-based build isolation behind a flag

pradyunsg opened the pull request:
This is still very much WIP, but it's been something I've been chipping away at for the past few weeks on a regular basis. I'm filing this eagerly, on the basis that "it can't hurt to file early".

I've incorporated some of the choices made in #11602 as a part of this; but that's triggered a bit of a rework of the existing changes I had.

Closes https://github.com/pypa/pip/issues/7778

pradyunsg committed:
Convert build_env into a sub-package

This will make it easier to organise the additional complexity of having
multiple build environment mechanisms.

pradyunsg committed:
Breakup the `build_env` module into smaller modules

This separates the various concerns internally, while still exposing the
same API for the sub-package.

pradyunsg committed:
Factor out the finder and logger dependent subprocess flags

This makes it possible to reuse this logic in places other than the
existing custom `BuildEnvironment` class.

pradyunsg committed:
Remove a no-longer-necessary indirection

This isn't particularly useful since the `_install_requirements` method
is no longer as complicated.

pradyunsg committed:
Remove an unused method

This method is not used or called by anything.

pradyunsg committed:
Implement a venv-based build environment

This enables using a virtual environment directly for handling the
creation of the build environment.

pradyunsg committed:
Rename `BuildEnvironment` to `CustomBuildEnvironment`

This makes room for a base / protocol for the build environment
interface and for alternative implementations.

pradyunsg committed:
Make `BuildEnvironment` a useful abstract base class

This moves shared logic into the base class for all build environments,
providing both runtime protections as well as type checker protections.

uranusjr commented on src/pip/_internal/build_env/_base.py:
```diff
@@ -0,0 +1,134 @@
+
+from pip._vendor.packaging.requirements import Requirement
+from pip._vendor.packaging.version import Version
+from pip._vendor.typing_extensions import Protocol
```
Gosh I never realised we’re already vendoring typing_extensions.

pradyunsg replied:
It was a fun one to realise that it's easy to add. :)

uranusjr commented on src/pip/_internal/build_env/_venv.py:
```diff
@@ -0,0 +1,106 @@
+        self._venv.create(self._temp_dir.path)
+
+        # Copy-pasted from venv/__init__.py
+        if sys.platform == "win32":
+            libpath = os.path.join(
+                self._temp_dir.path,
+                "Lib",
+                "site-packages",
+            )
+        else:
+            libpath = os.path.join(
+                self._temp_dir.path,
+                "lib",
+                f"python{sys.version_info.major}.{sys.version_info.minor}",
+                "site-packages",
+            )
```
Maybe we can use sysconfig here instead? New versions have a `venv` key for this specific use case, and we can fall back to the prefix scheme otherwise (or maybe limit this to 3.11+).

pradyunsg replied:
TBH, I'd like to switch over to this in a regular/slightly-long deprecation cycle to let us kill the old isolation logic and apply pressure on redistributors/users to unbreak their Python distributions. But, I digress.

Regarding sysconfig/venv scheme: yea, I think we could but this is matching the current implementation realities across all versions, and isn't particularly complex or difficult to maintain. It's only feasible for this to change in a new Python version which we will have the ability to adapt to. :)

pradyunsg commented:
BTW, @uranusjr let me know if you'd mind if I give you co-authorship on a bunch of these commits where I've taken out code from #11602 unchanged basically.

I've been a bit lazy in my commits here but I intend to do that.

pradyunsg committed:
Change `build_isolation` into a type-checked literal

This makes it possible for this to have multiple states while still
providing some level of protection via type-checking.

pradyunsg committed:
Raise a rich error when `venv` can't be imported

pradyunsg committed:
Plumb through `venv` build isolation

pradyunsg committed:
Pacify flake8 by re-exporting the variables under the same name

ichard26 commented:
@pradyunsg do you remember what work remained on this PR? After the pip 25.2 release (and once I get in-process build dependencies further along), I may be interested in picking this up.

ichard26 commented:
Superseded by PR https://github.com/pypa/pip/pull/14070.
